arm_gdb/mpu.py

```diff
@@ -0,0 +1,272 @@
+
+    return regions
+
+def get_rbar_reg():
+    return [
+        RegisterDef("MPU_RBAR", "MPU Region Base Address Register", 0xE000ED9C, 4, [
+            FieldBitfield("BASE", 5, 27, "Starting address of MPU region address"),
+            FieldBitfieldEnum("SH", 3, 2, [
+                (0b00, False, "Non-shareable", None),
+                (0b01, False, "Outer shareable", None),
+                (0b10, False, "Inner Shareable", None),
+                ],"Shareability for Normal memory"),
+            FieldBitfieldEnum("AP", 1, 2, [
+                (0b00, False, "read/write by privileged code only", None),
+                (0b01, False, "read/write by any privilege level", None),
+                (0b10, False, "Read only by privileged code only", None),
+                (0b11, False, "Read only by any privilege level", None),
+                ],"Access permissions"),
+            FieldBitfieldEnum("XN", 0, 1, [
+                (0b1, False, "Disallow program execution in this region", None),
+                (0b0, False, "Allow program execution in this region", None),
+                ],"eXecute Never attribute"),
+            ]),
+    ]
+
+def get_rlar_reg():
+    return [
+        RegisterDef("MPU_RLAR", "MPU Region Base Limit Register", 0xE000EDA0, 4, [
+            FieldBitfield("LIMIT", 5, 27, "Ending address (upper inclusive limit) of MPU region address"),
+            FieldBitfieldEnum("PXN", 4, 1, [
+                (0b1, False, "Execution from a privileged mode is not permitted", None),
+                (0b0, False, "Execution only permitted if read permitted", None),
+                ],"Privileged execute-never"),
+            FieldBitfield("AttrIndx", 1, 3, "Attribute Index. Select memory attributes from attribute sets in MPU_MAIR0 and MPU_MAIR1"),
+            FieldBitfieldEnum("EN", 0, 1, [
+                (0b0, False, "disable", None),
+                (0b1, False, "Region enable", None),
+            ],"Region enable"),
+        ]),
+    ]
+
+def get_mpu_region_start():
+    REG_RBAR = 0xE000ED9C
+    value_str = gdb.execute(f"x/1x {REG_RBAR}", to_string=True)
+    value_hex = value_str.split(':')[1].strip()
+    value = int(value_hex, 16)
+    start = value & 0xFFFFFFE0;
+    return start
+
+def get_mpu_region_limit():
+    REG_RLAR = 0xE000EDA0
+    value_str = gdb.execute(f"x/1x {REG_RLAR}", to_string=True)
+    value_hex = value_str.split(':')[1].strip()
+    value = int(value_hex, 16)
+    limit = value & 0xFFFFFFE0;
+    return limit
+
+def mpu_region_enable():
+    REG_RLAR = 0xE000EDA0
+    value_str = gdb.execute(f"x/1x {REG_RLAR}", to_string=True)
+    value_hex = value_str.split(':')[1].strip()
+    value = int(value_hex, 16)
+    enabled = bool(value & 0x01)
+    return enabled
+
+def get_mpu_regions():
+    set_rnr_reg(1)
+    reg = get_rbar_reg()
+    reg.dump(inf, args['descr'], base=base, all=args['all'])
+
+    reg = get_rlar_reg()
+    reg.dump(inf, args['descr'], base=base, all=args['all'])
```

Comment:
A getter (`get_*`) should have no side effects.

If something needs to be changed to access the correct data, then it needs to be reset so the state is not changed. Therefore, the `set_rnr_reg(...)` needs to be reset afterwards, if necessary.

But if I understand correctly, there are registers `MPU_RBAR_A1` and `MPU_RLAR_A1` that mirrors `MPU_RBAR` and `MPU_RLAR` for region 1, right? can't they be used instead of changing region?

Reply:
Even using MPU_RBAR_A1 & MPU_RLAR_A1, maybe still need the
change the MPU_RNR register, for there are only 4 alias registers, 
if the mpu number is large than 4(for one of my chip, there are 16 mpu regions).
| Condition  | When Accessing                  | MPU Region accessed |
|------------|----------------------------------|---------------------|
| MPU_RNR=0  | MPU_RBAR / MPU_RLAR              | 0                   |
|            | MPU_RBAR_A1 / MPU_RLAR_A1        | 1                   |
|            | MPU_RBAR_A2 / MPU_RLAR_A2        | 2                   |
|            | MPU_RBAR_A3 / MPU_RLAR_A3        | 3                   |
| MPU_RNR=4  | MPU_RBAR / MPU_RLAR              | 4                   |
|            | MPU_RBAR_A1 / MPU_RLAR_A1        | 5                   |
|            | MPU_RBAR_A2 / MPU_RLAR_A2        | 6                   |
|            | MPU_RBAR_A3 / MPU_RLAR_A3        | 7                   |